### PR TITLE
Yamauchi Coding Challenge

### PR DESCRIPTION
**Pull request by EduardoYamauchi**
Author: Eduardo Akimitsu Yamauchi.
Thanks for the challenge, was a great experience. 👋

Regards.

### CODE REVIEW

File: lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import 'package:hydrated_bloc/hydrated_bloc.dart';
+import 'package:restaurantour/models/restaurant.dart';
```
Comment (russell-sf) [inline lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_bloc.dart]: **suggestion**: Consistent import type.

File: lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_event.dart [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import '../../../models/restaurant.dart';
+
+abstract class FavoritesRestaurantsEvent {
+  const FavoritesRestaurantsEvent();
+}
+
+class AddedFavoriteRestaurant extends FavoritesRestaurantsEvent {
```
Comment (russell-sf) [inline lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_event.dart:7]: **suggestion**: Use freezed or equatable to eliminate boilerplate. Also using freezed for bloc event and state classes simplifies your code.

File: lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_state.dart [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import 'dart:convert';
+
+import '../../../models/restaurant.dart';
+
+class FavoritesRestaurantsState {
```
Comment (russell-sf) [inline lib/business_logic/bloc/favorites_restaurants/favorites_restaurants_state.dart:5]: **suggestion**: Use freezed or equatable to eliminate boilerplate. Also using freezed for bloc event and state classes simplifies your code.

File: lib/business_logic/bloc/restaurants/restaurants_state.dart [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import '../../../models/restaurant.dart';
+
+abstract class RestaurantsState {
```
Comment (russell-sf) [inline lib/business_logic/bloc/restaurants/restaurants_state.dart:3]: **suggestion**: Use freezed or equatable to eliminate boilerplate. Also using freezed for bloc event and state classes simplifies your code.

File: lib/business_logic/bloc/selected_restaurant/selected_restaurant_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import 'package:hydrated_bloc/hydrated_bloc.dart';
+import 'package:restaurantour/business_logic/bloc/selected_restaurant/selected_restaurant_event.dart';
+import 'package:restaurantour/business_logic/bloc/selected_restaurant/selected_restaurant_state.dart';
```
Comment (russell-sf) [inline lib/business_logic/bloc/selected_restaurant/selected_restaurant_bloc.dart]: **suggestion**: Use consistent import style.

File: lib/models/restaurant.dart [hunk 6]
```diff
@@ -16,6 +17,22 @@ class Category {
...
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Category &&
+          runtimeType == other.runtimeType &&
+          alias == other.alias &&
+          title == other.title;
+
+  @override
+  int get hashCode => alias.hashCode ^ title.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:35]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/models/restaurant.dart [hunk 7]
```diff
@@ -30,6 +47,19 @@ class Hours {
...
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Hours &&
+          runtimeType == other.runtimeType &&
+          isOpenNow == other.isOpenNow;
+
+  @override
+  int get hashCode => isOpenNow.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:62]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/models/restaurant.dart [hunk 8]
```diff
@@ -48,41 +78,89 @@ class User {
...
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is User &&
+          runtimeType == other.runtimeType &&
+          id == other.id &&
+          imageUrl == other.imageUrl &&
+          name == other.name;
+
+  @override
+  int get hashCode => id.hashCode ^ imageUrl.hashCode ^ name.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:92]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/models/restaurant.dart [hunk 9]
```diff
@@ -48,41 +78,89 @@ class User {
...
+      other is Review &&
+          runtimeType == other.runtimeType &&
+          id == other.id &&
+          rating == other.rating &&
+          text == other.text &&
+          user == other.user;
+
+  @override
+  int get hashCode =>
+      id.hashCode ^ rating.hashCode ^ text.hashCode ^ user.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:130]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/models/restaurant.dart [hunk 10]
```diff
@@ -48,41 +78,89 @@ class User {
...
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Location &&
+          runtimeType == other.runtimeType &&
+          formattedAddress == other.formattedAddress;
+
+  @override
+  int get hashCode => formattedAddress.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:160]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/models/restaurant.dart [hunk 11]
```diff
@@ -135,6 +213,70 @@ class Restaurant {
...
+      name: name ?? this.name,
+      price: price ?? this.price,
+      rating: rating ?? this.rating,
+      photos: photos ?? this.photos,
+      categories: categories ?? this.categories,
+      hours: hours ?? this.hours,
+      reviews: reviews ?? this.reviews,
+      location: location ?? this.location,
+    );
+  }
```
Comment (russell-sf) [inline lib/models/restaurant.dart:239]: **suggestion**: Freezed also provides this.

File: lib/models/restaurant.dart [hunk 12]
```diff
@@ -135,6 +213,70 @@ class Restaurant {
...
+  int get hashCode =>
+      id.hashCode ^
+      name.hashCode ^
+      price.hashCode ^
+      rating.hashCode ^
+      photos.hashCode ^
+      categories.hashCode ^
+      hours.hashCode ^
+      reviews.hashCode ^
+      location.hashCode;
```
Comment (russell-sf) [inline lib/models/restaurant.dart:279]: **suggestion**: Use freezed or equatable to eliminate boilerplate.

File: lib/presentation/widgets/divider_widget.dart [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+class DividerWidget extends StatelessWidget {
+  const DividerWidget({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) => Container(
+        color: Colors.grey.shade300,
+        height: 1,
+      );
+}
```
Comment (russell-sf) [inline lib/presentation/widgets/divider_widget.dart]: **question**: Why is this here? Flutter has a Divider widget.

**Review by russell-sf — COMMENTED** [review]
I looked over your submission and it looks good. Just a couple of patterns that I saw that you might want to improve.

* Imports. While there is no technical difference between relative and package imports within the same lib directory it is still a good idea to remain consistent.

* Boilerplate. You should not be writing methods such as copyWith, toString, operator == or hashCode. Those can be auto generated if you use a package such as Freezed.

You are using a package to generate your json code which is good!

Thanks for your submission.

**Comment by EduardoYamauchi** [comment]
> 
Thank you very much for your feedback. I really appreciate good and constructive feedback.

Imports: unfortunately I checked this twice, but I missed it. I use IntelliJ and when Intellij does an automatic refactor (especially when moving a file that uses other files as references), the refactor updates the references to 'package imports' instead of relative imports and I tend to use the relative import.

Boilerplate: I admit that for some reason I have avoided using generative builders for this project: 
- In the past, I have struggled with some low-level bugs caused by using the buildrunner lib, which has kept me from using generative code. (But I do not think that's the state of the art now, I think the lib is more stable).
- I avoided using packages to make code review easier and to keep the overhead as low as possible.

The point about error-proneness is very valid. I never thought about the benefits of using generated code.

I will make the changes. Thanks Russel!

File: lib/presentation/widgets/divider_widget.dart (hunk 13 above)
Comment (EduardoYamauchi) [reply lib/presentation/widgets/divider_widget.dart]: I really forget about the Divider Widget.

File: lib/presentation/screens/restarautour_home_screen.dart [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  static const titleText = 'RestauranTour';
+
+  static const tabBarAllRestaurantsText = 'All Restaurants';
+  static const tabBarMyFavoritesText = 'My Favorites';
+
+  static const appBarKey = Key('appBarKey');
+  static const tabBarKey = Key('tabBarKey');
+
+  @override
+  Widget build(BuildContext context) {
```
Comment (jshaw-mgm) [inline lib/presentation/screens/restarautour_home_screen.dart:22]: bug: The text is cut off.
Screen Shot 2022-08-09 at 8 43 17 AM

File: lib/presentation/screens/restaurant_details_screen.dart [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+            width: width,
+            height: width,
+            decoration: const BoxDecoration(color: Colors.white),
+            child: FadeInImage(
+              image: NetworkImage(restaurant.heroImage),
+              fit: BoxFit.cover,
+              placeholder: const AssetImage(
+                'assets/images/restaurant_placeholder.png',
+              ),
+              imageErrorBuilder: (_, __, ___) {
```
Comment (jshaw-mgm) [inline lib/presentation/screens/restaurant_details_screen.dart:50]: kudos: Awesome use of _ __ ___ !

File: lib/presentation/widgets/restaurant_list_tile.dart [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    fontWeight: FontWeight.w400,
+    wordSpacing: 1,
+    height: 1.5,
+  );
+
+  final Restaurant restaurant;
+
+  @override
+  Widget build(BuildContext context) {
+    final image = Padding(
```
Comment (jshaw-mgm) [inline lib/presentation/widgets/restaurant_list_tile.dart:31]: suggestion: The image is perfectly square but the designs show a rounded corner. 

Screen Shot 2022-08-09 at 8 48 13 AM

File: lib/presentation/widgets/restaurant_list_tile.dart [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        children: [
+          RatingWidget.simple(restaurant.rating),
+          RestaurantIsOpenWidget(restaurant),
+        ],
+      ),
+    );
+
+    final restaurantInformation = Padding(
+      padding: const EdgeInsets.fromLTRB(0, 9.3, 10, 5.15),
+      child: Stack(
```
Comment (jshaw-mgm) [inline lib/presentation/widgets/restaurant_list_tile.dart:84]: suggestion: Your layout here works but I don't find it easier to read or maintain using a stack of pre-made components.  This layout could be a bit improved for readability in my opinion.

File: test/restaurant_test.dart [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+void main() {
+  group('Restaurant Class', () {
+    Restaurant restaurant = TestFixture().restaurant();
+
+    test('Equals, hashCode and copyWithWith', () {
+      dynamic other;
+      Restaurant otherRestaurant = restaurant.copyWith();
+
+      // Reflexivity
```
Comment (jshaw-mgm) [inline test/restaurant_test.dart:16]: kudos: I appreciate the separation of tests with comments.  It makes it easy to see if everything was covered.

**Review by jshaw-mgm — COMMENTED** [review]
Thank you for taking the time to submit the code for review.  I know how challenging and time consuming this can be and I am grateful for your time and willingness to demonstrate your knowledge of Flutter.  

I have left a few comments.  Please don't feel like you need to make any changes and resubmit. 

Overall I thought that the code was well organized and easy to follow what you did.

File: lib/presentation/widgets/restaurant_list_tile.dart (hunk 16 above)
Comment (EduardoYamauchi) [reply lib/presentation/widgets/restaurant_list_tile.dart:31]: Unfortunately, this is an error that I missed. I always check the design proposal: corners, fonts, transitions, image size, typography, etc. The design is one of the things that add value to the application.

I think I removed the corner when I implemented the placeholder image when loading from the internet.

File: lib/presentation/screens/restarautour_home_screen.dart (hunk 14 above)
Comment (EduardoYamauchi) [reply lib/presentation/screens/restarautour_home_screen.dart:22]: This is a bug. When I used the emulator with the default iPhone 13, I didn't notice that. I switch to a different screen size/resolution and it appears.

File: lib/presentation/widgets/restaurant_list_tile.dart (hunk 17 above)
Comment (EduardoYamauchi) [reply lib/presentation/widgets/restaurant_list_tile.dart:84]: Yes, I agree. Separating this UI logic in a widget could be a start good solution.

**Comment by EduardoYamauchi** [comment]
> Thank you for taking the time to submit the code for review. I know how challenging and time consuming this can be and I am grateful for your time and willingness to demonstrate your knowledge of Flutter.
> 
> I have left a few comments. Please don't feel like you need to make any changes and resubmit.
> 
> Overall I thought that the code was well organized and easy to follow what you did.

You both have paid a lot of attention to reviewing this code, I really appreciate that. It's nice to know that Superformula dedicate a good effort to ensure the quality of the code. Thank you for every feedback!
